File: platform/windows/TailScout.Windows.Core/Models/TailscaleProfile.cs
namespace TailScout.Windows.Core.Models;

public sealed record TailscaleProfile(
    string Id,
    string Nickname,
    string Tailnet,
    string Account,
    bool Selected)
{
    public string DisplayName => Text.First(Nickname, Account, Tailnet, Id);
    public string SwitchKey => Text.First(Id, Nickname, Account, Tailnet);

    public static IReadOnlyList<TailscaleProfile> ParseList(string json) =>
        JsonModel.Parse<RawProfile[]>(json)
            .Select(profile => new TailscaleProfile(
                Text.Value(profile.Id), Text.Value(profile.Nickname), Text.Value(profile.Tailnet),
                Text.Value(profile.Account), profile.Selected ?? false))
            .Where(profile => profile.DisplayName.Length > 0)
            .ToArray();
}

internal sealed class RawProfile
{
    public string? Id { get; init; }
    public string? Nickname { get; init; }
    public string? Tailnet { get; init; }
    public string? Account { get; init; }
    public bool? Selected { get; init; }
}

File: platform/macos/Sources/TailScoutCore/TailscaleProfile.swift
import Foundation

public struct TailscaleProfile: Decodable, Equatable, Sendable {
    public let id: String
    public let nickname: String
    public let tailnet: String
    public let account: String
    public let selected: Bool

    enum CodingKeys: String, CodingKey {
        case id, nickname, tailnet, account, selected
    }

    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        id = try values.decodeString(forKey: .id)
        nickname = try values.decodeString(forKey: .nickname)
        tailnet = try values.decodeString(forKey: .tailnet)
        account = try values.decodeString(forKey: .account)
        selected = try values.decodeBool(forKey: .selected)
    }

    public static func parseList(_ input: String) throws -> [Self] {
        try JSONDecoder().decode([Self].self, from: Data(input.utf8))
            .filter { !$0.displayName.isEmpty }
    }

    public var displayName: String {
        [nickname, account, tailnet, id].first(where: { !$0.isEmpty }) ?? ""
    }

    public var switchKey: String {
        [id, nickname, account, tailnet].first(where: { !$0.isEmpty }) ?? ""
    }
}
